Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+                while self.receivedMessages.count > 0 {
+                    let bufferedPart = self.receivedMessages.removeFirst()
+                    context.fireChannelRead(bufferedPart)
+                }
+                if fireReadComplete {
+                    context.fireChannelReadComplete()
+                }
+
+            }
+            .whenComplete { _ in
+                context.pipeline.removeHandler(context: context, promise: nil)
+            }
+        }
+    }
+    
+    /// Removes any extra HTTP-related handlers from the channel pipeline.
+    private func removeHTTPHandlers(context: ChannelHandlerContext) -> EventLoopFuture<Void> {
+        guard self.httpHandlers.count > 0 else {
+            return context.eventLoop.makeSucceededFuture(())
+        }
+        
+        let removeFutures = self.httpHandlers.map { context.pipeline.removeHandler($0) }
+        return .andAllSucceed(removeFutures, on: context.eventLoop)
+    }
+    
+    private func gotUpgrader(upgrader: @escaping (() -> Void)) {
+
+        guard case .awaitingUpgrader = self.upgradeState else {
+            preconditionFailure("Unexpected upgrader state: \(self.upgradeState)")
+        }
+
+        self.upgradeState = .upgraderReady(upgrader)
+        if self.seenFirstResponse {
+            // Ok, we're good to go, we can upgrade. Otherwise we're waiting for .end, which
+            // will trigger the upgrade.
+            upgrader()
+        }
+    }
+    
+    private func notUpgrading(context: ChannelHandlerContext, data: HTTPClientResponsePart) {
+        self.upgradeState = .upgradeFailed
+        
+        if !self.seenFirstResponse {
+            // We haven't seen the first response .end. That means we're not buffering anything, and we can
+            // just deliver data.
+            assert(self.receivedMessages.count == 0)
+            context.fireChannelRead(self.wrapInboundOut(data))
+        } else {
+            // We've seen the first request .end, so we now need to deliver the .head we got passed,
+            // as well as the .end we swallowed, and any buffered parts. While we're doing this
+            // we may be reentrantly called, which will cause us to buffer new parts. To make that safe, we
+            // must ensure we aren't holding the buffer mutably, so no for loop for us.
+            context.fireChannelRead(self.wrapInboundOut(data))
+            context.fireChannelRead(self.wrapInboundOut(.end(nil)))
+            
+            while self.receivedMessages.count > 0 {
+                let bufferedPart = self.receivedMessages.removeFirst()
+                context.fireChannelRead(bufferedPart)
+            }
+        }
+        
+        // Ok, we've delivered all the parts. We can now remove ourselves, which should happen synchronously.
+        context.pipeline.removeHandler(context: context, promise: nil)
+    }
+    
+    public func errorCaught(context: ChannelHandlerContext, error: Error) {
```

Comment:
Why is this handler taking it upon itself to close on error?

Reply:
This is a mistake and is just something I accidentally left in from a now defunct previous implementation. Good spot. Removed.